feat: ability to show UI Bar values as percentage

Arufonsu opened the pull request:
- chore: some things were reorganized around the modified code.
- new feat: ability to show UI Bar values as percentage.
- If enabled, will display the toggled bar's as percentage and it's tooltip will show the actual value instead.
- If disabled, will display the toggled bar's actual value and it's tooltip will show the percentage instead.
- should resolve #1431 

**Assets PR :** https://github.com/AscensionGameDev/Intersect-Assets/pull/22

### Preview

https://user-images.githubusercontent.com/17498701/189506575-57de6696-52cf-4656-be1c-4428e3d7e366.mp4

lodicolo commented on Intersect.Client.Framework/Database/GameDatabase.cs:
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
+
+        public bool PlayerOverheadInfo { get; set; }
+
+        public bool HideOthersOnWindowOpen { get; set; }
```
Why was this moved? It made reading this diff harder than necessary.

Arufonsu replied:
My bad, at some point i meant to re-organize these variables according to their categories with comments, let me know if i should undo this one. as for the other properties, I've just added their missing `{ get; set; }`

Arufonsu replied:
moved back `HideOthersOnWindowOpen`

lodicolo commented on Intersect.Client.Framework/Database/GameDatabase.cs:
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
+        public bool UiExpToPercentage { get; set; }
 
-        public bool PartyMemberOverheadInfo;
+        public bool UiHpToPercentage { get; set; }
```
`ShowHealthAsPercentage`

lodicolo commented on Intersect.Client.Framework/Database/GameDatabase.cs:
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
+        public bool HideOthersOnWindowOpen { get; set; }
 
-        public bool NpcOverheadInfo;
+        public bool UiExpToPercentage { get; set; }
```
`ShowExperienceAsPercentage`

lodicolo commented on Intersect.Client.Framework/Database/GameDatabase.cs:
```diff
@@ -26,17 +23,25 @@ public abstract partial class GameDatabase
+        public bool UiHpToPercentage { get; set; }
 
-        public bool PlayerOverheadInfo;
+        public bool UiMpToPercentage { get; set; }
```
`ShowManaAsPercentage`

lodicolo commented on Intersect.Client.Framework/Database/GameDatabase.cs:
```diff
@@ -76,6 +81,9 @@ public virtual void LoadPreferences()
             NpcOverheadInfo = LoadPreference("NpcOverheadInfo", true);
             PartyMemberOverheadInfo = LoadPreference("PartyMemberOverheadInfo", true);
             PlayerOverheadInfo = LoadPreference("PlayerOverheadInfo", true);
+            UiExpToPercentage = LoadPreference("UiExpToPercentage", true);
+            UiHpToPercentage = LoadPreference("UiHpToPercentage", false);
+            UiMpToPercentage = LoadPreference("UiMpToPercentage", false);
```
Use `nameof` instead of a string literal

lodicolo commented on Intersect.Client.Framework/Database/GameDatabase.cs:
```diff
@@ -95,6 +103,9 @@ public virtual void SavePreferences()
             SavePreference("NpcOverheadInfo", NpcOverheadInfo.ToString());
             SavePreference("PartyMemberOverheadInfo", PartyMemberOverheadInfo.ToString());
             SavePreference("PlayerOverheadInfo", PlayerOverheadInfo.ToString());
+            SavePreference("UiExpToPercentage", UiExpToPercentage.ToString());
+            SavePreference("UiHpToPercentage", UiHpToPercentage.ToString());
+            SavePreference("UiMpToPercentage", UiMpToPercentage.ToString());
```
Use `nameof` instead of a string literal

Arufonsu replied:
like this? :

![image](https://user-images.githubusercontent.com/17498701/189537274-bdd64dfc-bd03-4cf4-9f5e-49d35836e22e.png)

should i change the rest within `SavePreferences()` as well?

lodicolo commented on Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:
```diff
@@ -585,10 +585,24 @@ private void UpdateHpBar(float elapsedTime, bool instant = false)
-                HpLbl.Text = Strings.EntityBox.vital0val.ToString(
-                    MyEntity.Vital[(int) Vitals.Health], MyEntity.MaxVital[(int) Vitals.Health]
+                float hpPercentage = hpfillRatio * 100;
+                var hpPercentageText = hpPercentage == 0 ? $"{hpPercentage}%" : $"{hpPercentage:0.0}%";
```
This ternary is unnecessary, use `{hpPercentage:0.##}%` instead.

lodicolo commented on Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:
```diff
@@ -684,14 +698,30 @@ private void UpdateHpBar(float elapsedTime, bool instant = false)
-                MpLbl.Text = Strings.EntityBox.vital1val.ToString(
-                    MyEntity.Vital[(int) Vitals.Mana], MyEntity.MaxVital[(int) Vitals.Mana]
+                float mpPercentage = targetMpWidth * 100;
+                var mpPercentageText = mpPercentage == 0 ? $"{mpPercentage}%" : $"{mpPercentage:0.0}%";
```
This ternary is unnecessary, use `{mpPercentage:0.##}%` instead.

lodicolo commented on Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:
```diff
@@ -742,14 +772,29 @@ private void UpdateMpBar(float elapsedTime, bool instant = false)
+                targetExpWidth = (float)((Player)MyEntity).Experience /
+                                 (float)((Player)MyEntity).GetNextLevelExperience();
+                float expPercentage = targetExpWidth * 100;
+                var expPercentageText = expPercentage == 0 ? $"{expPercentage}%" : $"{expPercentage:0.00}%";
```
This ternary is unnecessary, use `{expPercentage:0.##}%` instead.

lodicolo commented on Intersect.Client/Interface/Shared/SettingsWindow.cs:
```diff
@@ -193,17 +199,25 @@ public SettingsWindow(Base parent, MainMenu mainMenu, EscapeMenu escapeMenu)
             mAutoCloseWindowsCheckbox.Text = Strings.Settings.AutoCloseWindows;
 
+            // Game Settings - Toggle for: User Interface EXP/HP/MP to Percentage.
+            mUiExpToPercentageCheckbox = new LabeledCheckBox(mGameSettingsContainer, "UiExpToPercentageCheckbox");
+            mUiExpToPercentageCheckbox.Text = Strings.Settings.UiExpToPercentage;
+            mUiHpToPercentageCheckbox = new LabeledCheckBox(mGameSettingsContainer, "UiHpToPercentageCheckbox");
+            mUiHpToPercentageCheckbox.Text = Strings.Settings.UiHpToPercentage;
+            mUiMpToPercentageCheckbox = new LabeledCheckBox(mGameSettingsContainer, "UiMpToPercentageCheckbox");
+            mUiMpToPercentageCheckbox.Text = Strings.Settings.UiMpToPercentage;
+            
```
The `Ui` prefix is not beneficial here at all, it's just two extra characters.

lodicolo commented on Intersect.Client/Localization/Strings.cs:
```diff
@@ -1787,6 +1787,15 @@ public partial struct Settings
             public static LocalizedString Vsync = @"V-Sync";
+            
+            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+            public static LocalizedString UiExpToPercentage = @"EXP to Percentage";
```
`Show Experience as percentage`

Arufonsu replied:
`Show Experience as percentage` it too long and hard to fit in the graphic image-box that belongs to the interface settings ; would you agree with `Experience as percentage` instead?

lodicolo commented on Intersect.Client/Localization/Strings.cs:
```diff
@@ -1787,6 +1787,15 @@ public partial struct Settings
+            public static LocalizedString UiExpToPercentage = @"EXP to Percentage";
+            
+            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+            public static LocalizedString UiHpToPercentage = @"HP to Percentage";
```
`Show Health as percentage`

Arufonsu replied:
https://github.com/AscensionGameDev/Intersect-Engine/pull/1448/files#r967848965

lodicolo commented on Intersect.Client/Localization/Strings.cs:
```diff
@@ -1787,6 +1787,15 @@ public partial struct Settings
+            public static LocalizedString UiHpToPercentage = @"HP to Percentage";
+            
+            [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
+            public static LocalizedString UiMpToPercentage = @"MP to Percentage";
```
`Show Mana as percentage`

Arufonsu replied:
https://github.com/AscensionGameDev/Intersect-Engine/pull/1448/files#r967848965

lodicolo commented on Intersect.Client/Interface/Game/EntityPanel/EntityBox.cs:
```diff
@@ -742,14 +772,29 @@ private void UpdateMpBar(float elapsedTime, bool instant = false)
+                else
+                {
+                    ExpLbl.Text = expValueText;
+                    ExpBackground.SetToolTipText(expPercentageText);
+                    ExpBar.SetToolTipText(expPercentageText);
```
Is setting both tooltips really necessary? Is there a difference in behavior if only one of them is set?

Arufonsu replied:
I was doing this because bars and their backgrounds had `"MouseInputEnabled": true` , so if i added the tooltip to the background only, the bar wouldn't allow it to popup when being over the background.
today i've just realized that having the tooltip enabled for the background is more than enough, and in order for this to work i'll have to change the assets so the actual bars over the background have `MouseInputEnabled` disabled.
I'll change this in the next commit.

Arufonsu commented:
Updated with review, [also updated the assets](https://github.com/AscensionGameDev/Intersect-Assets/pull/22) due to some changes.

Preview:

https://user-images.githubusercontent.com/17498701/189541998-d93331f3-e59d-4c0f-bfb6-f52688689db5.mp4

Arufonsu commented:
Updated with panda's review, [also updated the assets](https://github.com/AscensionGameDev/Intersect-Assets/pull/22) due to some changes.

Preview:

![image](https://user-images.githubusercontent.com/17498701/189760702-e5a2b1ff-a02d-4700-b204-bde16be134eb.png)
